### PR TITLE
feat: Hooks added to the "OrdersTableQuery::build_query()" function

### PR DESCRIPTION
**Pull request by kidunot89**
### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

- Recreation of #38595 
Adds the following hooks to the **Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableQuery** class.
- `woocommerce_cot_query_join`: Meant to work like the `posts_join` hook but for the orders table.
- `woocommerce_cot_query_where`: Meant to work like the `posts_where` hook but for the orders table.
- `woocommerce_cot_query_orderby`: Meant to work like the `posts_orderby` hook but for the orders table.
- `woocommerce_cot_query_groupby`: Meant to work like the `posts_groupby` hook but for the orders table.

No none issues related to this. I personally just need this functionality to fully support HPOS is [WPGraphQL WooCommerce](https://github.com/wp-graphql/wp-graphql-woocommerce)

### How to test the changes in this Pull Request:

Testing these hooks effectively can get a complicated so below is a test plugin.

```php
<?php
/**
 * Test plugin
 */

/**
 * Edit the WHERE clause of the query.
 *
 * @param string $where The WHERE clause of the query.
 * @param array  $args  The arguments passed to the query.
 * @param string $query The query being filtered.
 *
 * @return string
 */
function edit_where( $where, $args, $query ) {
    error_log( $where );

    return $where;
}
add_filter( 'woocommerce_cot_query_where', 'edit_where', 10, 3 );
        

/**
 * Edit the ORDER BY clause of the query.
 *
 * @param string $orderby The orderby clause of the query.
 * @param array  $args    The arguments passed to the query.
 * @param string $query   The query being filtered.
 *
 * @return string
 */
function edit_orderby( $orderby, $args, $query ) {
    error_log( $orderby );

    return $orderby;
}
add_filter( 'woocommerce_cot_query_orderby', 'edit_orderby', 10, 3 );
```
1. Set WP_DEBUG_LOG flag.
2. Install and activate WooCommerce and the provided test plugin
3. Activate HPOS from the WC settings.
4. Navigate to `Orders` page on the admin.
5. Check debug log of output of "where" and "orderby"

### CODE REVIEW

**Comment by coreymckrill** [comment]
@barryhughes Stealing this one from you because I was assigned to the [original](https://github.com/woocommerce/woocommerce/pull/38595).

**Comment by barryhughes** [comment]
That's ok 😢

File: pnpm-lock.yaml
Comment (coreymckrill) [inline pnpm-lock.yaml:1]: 🤔 Not sure how this got in here, but these changes should be reverted and this file left out of the changeset...

**Review by coreymckrill — COMMENTED** [review]
👍 It's looking good, the test code outputs the right things. Just a couple of minor things for this first round. I'm still checking on a couple of other things, but can hopefully wrap those up tomorrow.

The phpcs linting rules would really like it if the doc blocks for each of the new filters had a `@since` parameter. You can put `7.9.0` for the version for now.

**Comment by kidunot89** [comment]
@coreymckrill Done and done.

**Comment by coreymckrill** [comment]
@kidunot89 you mentioned these hooks are necessary to make WooGraphQL compatible with HPOS. Could you provide an example or two of how these would be used?

**Comment by coreymckrill** [comment]
@kidunot89 It has been brought to my attention that there's another PR already that does something very similar to this one: https://github.com/woocommerce/woocommerce/pull/36625

Since that other one is further along and also has unit tests, we might go with that one instead. Would it be sufficient for the compatibility work you're doing with WooGraphQL?

**Comment by kidunot89** [comment]
Yep, just reviewed the code and it looks like it should work.

**Comment by kidunot89** [comment]
@coreymckrill I'll be closing this now then

**Pull request closed without merging**